src/AzureStorageQueueWorker.cs

```diff
@@ -0,0 +1,332 @@
+        {
+            var processorName = typeof(TMessageProcessor).FullName;
+            using var logScope = _logger.BeginScope("{Processor} reads {queueName}", processorName, _workerOptions.QueueName);
+            try
+            {
+                var queueClient = new QueueClient(_workerOptions.AzureStorageConnectionString, _workerOptions.QueueName);
+                await queueClient.CreateIfNotExistsAsync();
+                while (!stoppingToken.IsCancellationRequested)
+                {
+                    var processed = false;
+                    MessageDetails messageDetails = null;
+                    QueueMessage queueMessage = null;
+                    CancellationTokenSource updateMessageTokenSource = null;
+                    Task updateQueueMessageTask = null;
+                    IServiceScope scope = null;
+                    try
+                    {
+                        var queueMessageResponse = await queueClient.ReceiveMessagesAsync(1, _workerOptions.VisibilityTimeout, stoppingToken);
+                        queueMessage = queueMessageResponse.Value.FirstOrDefault();
+
+                        if (queueMessage == null)
+                        {
+                            _logger.LogInformation("No messages, sleeping...");
+                            await Task.Delay(_workerOptions.NoMessagesSleepDuration, stoppingToken);
+                            continue;
+                        }
+
+                        //TODO: we need more metric. This is just very early version to check. How it'll work
+                        using var tc = _telemetryClient.StartOperation<DependencyTelemetry>($"{processorName} queue: {_workerOptions.QueueName}");
+                        messageDetails = new MessageDetails
+                        {
+                            Message = queueMessage,
+                            Receipt = queueMessage.PopReceipt
+                        };
+
+                        updateMessageTokenSource = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
+                        var visibilityToken = updateMessageTokenSource.Token;
+                        updateQueueMessageTask = Task.Run(() =>
+                            KeepMessageInvisibleAsync(queueClient, messageDetails, visibilityToken), visibilityToken);
+
+                        scope = _serviceProvider.CreateScope();
+                        var messageProcessor = scope.ServiceProvider.GetRequiredService<TMessageProcessor>();
+                        await messageProcessor.ProcessMessage(GetTypedMessage(queueMessage), stoppingToken);
+                        processed = true;
+
+                        //stoppingToken aren't passed here. We should try to delete message even if cancellation was requested
+                        //if operation was done
+                        await DeleteMessageAsync(queueClient, messageDetails, updateMessageTokenSource);
```

Comment:
Won't `updateMessageTokenSource` be cancelled whenever `stoppingToken` gets cancelled?

Reply:
yes, `updateMessageTokenSource.Token` will become cancelled. But `updateMessageTokenSource` is passed to cancel `UpdateVisibility` logic on message deletion

Reply:
ahh, yea - makes sense